File: Level 1/[카카오 인턴] 키패드 누르기/solution.py
def solution(numbers, hand):
    keypad = {
        1: (0, 0),
        2: (0, 1),
        3: (0, 2),
        4: (1, 0),
        5: (1, 1),
        6: (1, 2),
        7: (2, 0),
        8: (2, 1),
        9: (2, 2),
        "*": (3, 0),
        0: (3, 1),
        "#": (3, 2),
    }

    left = "*"
    right = "#"

    answer = ""

    for num in numbers:
        if num in [1, 4, 7]:
            left = num
            answer += "L"
        elif num in [3, 6, 9]:
            right = num
            answer += "R"
        else:
            left_distance = abs(keypad[num][0] - keypad[left][0]) + abs(
                keypad[num][1] - keypad[left][1]
            )
            right_distance = abs(keypad[num][0] - keypad[right][0]) + abs(
                keypad[num][1] - keypad[right][1]
            )
            if left_distance == right_distance:
                if hand == "right":
                    right = num
                    answer += "R"
                else:
                    left = num
                    answer += "L"
            elif left_distance < right_distance:
                left = num
                answer += "L"
            else:
                right = num
                answer += "R"

    return answer

File: Level 1/[카카오 인턴] 키패드 누르기/solution.js
function solution(numbers, hand) {
  const keypad = {
    1: [0, 0],
    2: [1, 0],
    3: [2, 0],
    4: [0, 1],
    5: [1, 1],
    6: [2, 1],
    7: [0, 2],
    8: [1, 2],
    9: [2, 2],
    "*": [0, 3],
    0: [1, 3],
    "#": [2, 3],
  };

  let left = "*";
  let right = "#";

  return numbers
    .map((num) => {
      if ([1, 4, 7].includes(num)) {
        left = num;
        return "L";
      }

      if ([3, 6, 9].includes(num)) {
        right = num;
        return "R";
      }

      const leftDistance =
        Math.abs(keypad[num][0] - keypad[left][0]) +
        Math.abs(keypad[num][1] - keypad[left][1]);
      const rightDistance =
        Math.abs(keypad[num][0] - keypad[right][0]) +
        Math.abs(keypad[num][1] - keypad[right][1]);

      if (leftDistance === rightDistance) {
        if (hand === "right") {
          right = num;
          return "R";
        } else {
          left = num;
          return "L";
        }
      }

      if (leftDistance < rightDistance) {
        left = num;
        return "L";
      } else {
        right = num;
        return "R";
      }
    })
    .join("");
}
